src/Compiler/Checking/CheckExpressions.fs

```diff
@@ -9713,6 +9731,37 @@ and TcMethodApplication
     // Handle post-hoc property assignments
     let setterExprPrebinders, callExpr3 =
         let expr = callExpr2b
+
+        // Make sure, if apparent type has any required properties, they all are in the `finalAssignedItemSetters`.
+        // If if is a constructor, and it is not marked with `SetsRequiredMembersAttributeAttribute`, then:
+        // 1. Get all properties of the type.
+        // 2. Check if any of them has `IsRequired` set.
+        //      2.1. If there are none, proceed as usual
+        //      2.2. If there are any, make sure all of them (or their setters) are in `finalAssignedItemSetters`.
+        // 3. If some are missing, produce a diagnostic which missing ones.
+        if g.langVersion.SupportsFeature(LanguageFeature.RequiredPropertiesSupport) && finalCalledMethInfo.IsConstructor then
+            let requiredProps = 
+                match finalCalledMethInfo with
+                | ILMeth (_, ilmethInfo, _) ->
+                    if TryFindILAttribute g.attrib_SetsRequiredMembersAttribute ilmethInfo.RawMetadata.CustomAttrs then
+                        []
+                    else     
+                        let props = GetImmediateIntrinsicPropInfosOfType (None, AccessibleFromSomeFSharpCode) g cenv.amap range0 ilmethInfo.ApparentEnclosingType
+                        List.filter (fun (p: PropInfo) -> p.IsRequired ) props
+                | _ -> []
+            
+            if requiredProps.Length > 0 then
+                let setterPropNames = 
```

Comment:
Not entirely sure we should check only via names here, @dsyme would that be proper way of checking for missing props?